### DISCUSSION TITLE
Howto replace docker base image when using latest tag?

### DISCUSSION THREAD

**Discussion by jannylund** [renovatebot/renovate#19595] [Request Help]
We have a use case where quite many of our internal applications are built on top of the same docker base image, let's call it

`registry.company.com/oldbase:latest`

Now, since we are moving towards new base images, we were planning to let renovate do the heavy work and replace this with:

`registry.company.com/newbase:latest`

To do this, I tried a few version of config, but in essence, this is what I assumed would do the job. (possibly while also setting `matchCurrentVersion` and `replacementVersion` to `latest`

```
    {
      "matchPackageNames": [
        "registry.company.com/oldbase"
      ],
      "replacementName": "registry.company.com/newbase",
    }
```

However, it seems I never get to this step, as Renovate throws me a warning
```
build	28-Dec-2022 14:35:26	DEBUG: Dependency registry.company.com/oldbase has unsupported value latest (repository=mysamplerepo)
```

And then this additional comment about `invalid-value`

```
build	28-Dec-2022 14:35:28	DEBUG: packageFiles with updates (repository=mysamplerepo, baseBranch=master)
build	28-Dec-2022 14:35:28	       "config": {
build	28-Dec-2022 14:35:28	         "dockerfile": [
build	28-Dec-2022 14:35:28	           {
build	28-Dec-2022 14:35:28	             "packageFile": "Dockerfile",
build	28-Dec-2022 14:35:28	             "deps": [
build	28-Dec-2022 14:35:28	               {
build	28-Dec-2022 14:35:28	                 "depName": "registry.company.com/oldbase",
build	28-Dec-2022 14:35:28	                 "currentValue": "latest",
build	28-Dec-2022 14:35:28	                 "replaceString": "registry.company.com/oldbase:latest",
build	28-Dec-2022 14:35:28	                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
build	28-Dec-2022 14:35:28	                 "datasource": "docker",
build	28-Dec-2022 14:35:28	                 "depType": "final",
build	28-Dec-2022 14:35:28	                 "depIndex": 0,
build	28-Dec-2022 14:35:28	                 "updates": [],
build	28-Dec-2022 14:35:28	                 "warnings": [],
build	28-Dec-2022 14:35:28	                 "versioning": "docker",
build	28-Dec-2022 14:35:28	                 "skipReason": "invalid-value"
build	28-Dec-2022 14:35:28	               }
build	28-Dec-2022 14:35:28	             ]
build	28-Dec-2022 14:35:28	           }
build	28-Dec-2022 14:35:28	         ],
```

Is there something I'm missing here, or is this currently just not supported with the `latest` tag? 🤔 

I would assume this would be a nice feature overall, to also allow migrating from `latest` to the actual pinned version.

**Comment by WhoAmI0501** [comment]
Hi,
as [the documentation states](https://docs.renovatebot.com/configuration-options/#replacementversion), `replacementVersion` is only supported by the `npm` manager. So you cannot use `replacementVersion` for your `Dockerfile`.

I am currently also struggeling with the `gitlabci-include` manager. I do not have a solution, but: The value `latest` does not match the `docker` versioning scheme. So you must write a own Regex-Versioning like described [here](https://docs.renovatebot.com/modules/versioning/#regular-expression-versioning), to match the latest pattern.

This doesnt replace the string, but it may helps you...

**Reply by jannylund** [reply]
Thanks, but looks like this is not possible, as the regex versioning requires at least one of major/minor/patch.

I had a look at the codebase and seems `ivy` versioning supports the `latest` tag. Setting versioning to ivy will however make Renovate try to lookup the tags for `oldbase` which fails. I would have expected a lookup instead towards the `newbase` 🤔 

```
build	28-Dec-2022 23:12:00	DEBUG: Page Not Found https://registry.company.com/v2/oldbase/tags/list?n=10000 (repository=mysamplerepo)
build	28-Dec-2022 23:12:00	DEBUG: Retrying Tags for https://registry.company.com/oldbase using library/ prefix (repository=mysamplerepo)
```

**Reply by jannylund** [reply]
Ok, found a bug in an nginx conf and the lookups are now working. Still, the problem is that Renovate will not update anything when both the source and the target has the same version, that is, the tag `latest`.

I also gave regexManager a shot but seems it fails on the same idea, it will simply not trigger as it does not see any reason to change from `latest` to `latest` and therefore it also does not replace the image name.

Thinking that this seems to be a dead end for now. An optimal solution would actually be if renovate would have a feature to simply define a regex matcher that does a string replacement and no lookups. 🤔

**Reply by jannylund** [reply]
The `replacementVersion` feature is now available for Docker, but since `latest` is filtered out of docker tags, it's still not possible to do what I'm trying to do. It actually seems kinda like a bug, that in the corner case of setting another versioning strategy, it is not respected when checking available versions from the source. So even when versioning i set to `ivy` (which allows `latest`), the source version check will use docker versioning and filter out `latest`.

I wonder if it would make sense to:
- implement support for allowing the set `replacementVersion` to override what is returned from the versions listed by renovate.
- implement support for making the versioning respect the versioning that is set.
- implement a switch to allow `latest` as valid version for docker. I assume this should be off by default to not mess with the current functionality that simply does not allow `latest` as a tag.

I would be happy to give the implementation a shot, but would be good to understand what others need, and if there is any idea even to try getting such a thing merged. 🤔

**Reply by JamieMagee** [reply]
Yeah, it's definitely a corner case right at the minute. And not just for Docker either. #19805 solves the case for invalid versions, but there might be more required on top, to update the 'latest' version.
